ts-apps/decision-logic-visualizer/src/lib/algebraic-graphs/dag.ts

```diff
@@ -77,37 +77,60 @@ export abstract class Dag<A extends Ord<A> & HasId>
       .with([], () => empty<A>())
       .otherwise(() => vertex(topSort[topSort.length - 1]))
   }
+
+  /** Internal helper */
+  protected getAllPathsFromVertex(vertex: A): Array<Array<A>> {
+    const neighbors = this.getAdjMap().get(vertex) ?? new Set()
+    if (neighbors.size === 0) return [[vertex]]
+
+    const pathsFromNeighbors = Array.from(neighbors).flatMap((neighbor) =>
+      this.getAllPathsFromVertex(neighbor)
+    )
+    return pathsFromNeighbors.map((path) => [vertex, ...path])
+  }
+
+  // TODO: The ret type should be Dag[], so tt
+  // it's easier to preserve metadata associated with the edges
+  getAllPaths() {
+    const source = this.getSource()
+    const barePaths: Array<Array<A>> = match(source)
+      .with(P.instanceOf(Empty), () => [])
+      .otherwise((source) =>
+        this.getAllPathsFromVertex((source as Vertex<A>).getValue())
+      )
+    return barePaths.map(pathFromValues)
+  }
 }
 
 /*********************
      Primitives
 **********************/
 
-export function isEmpty<A extends Ord<A> & HasId>(g: DirectedAcyclicGraph<A>) {
+export function isEmpty<A extends Ord<A>>(g: DirectedAcyclicGraph<A>) {
   return g instanceof Empty
 }
 
-export function empty<A extends Ord<A> & HasId>() {
+export function empty<A extends Ord<A>>() {
   return new Empty<A>()
 }
 
 /** Empty graph */
-export class Empty<A extends Ord<A> & HasId> extends Dag<A> {
+export class Empty<A extends Ord<A>> extends Dag<A> {
   constructor() {
     super()
   }
 }
 
-export function isVertex<A extends Ord<A> & HasId>(g: DirectedAcyclicGraph<A>) {
+export function isVertex<A extends Ord<A>>(g: DirectedAcyclicGraph<A>) {
```

Comment:
```suggestion
export function isVertex<A extends Ord<A>>(g: DirectedAcyclicGraph<A>): g is Vertex<A> {
```
Then it becomes a proof that allows you to narrow down the type of `g`.

Reply:
Great catch, thank you!